### DISCUSSION TITLE
Ability to adjust pagefind search result weighting on a page level

### DISCUSSION THREAD

**Discussion by lorenzolewis** [withastro/starlight#2348] [Feature Requests]
### What version of `starlight` are you using?

0.28.1

### What is your idea?

Right now the only customization we have for Pagefind results is to include or exclude a page from search results (https://starlight.astro.build/guides/site-search/#exclude-a-page).

Additionally, you can directly modify the ranking of a specific part of a page using Pagefind's ranking mechanism: https://pagefind.app/docs/weighting/. However, this can be a bit awkward when doing inside of markdown files.

Right now there is no way to change the overall weight or ranking of a single page in the context of the returned search results within going into component overrides or hard-coding some HTML within your markdown files.

We could extend the `pagefind` front matter like this:

```md
---
title: Important doc
pagefind:
  rank: 10 // any number between 0.0 and 10.0
---
```

The rank can be any number between 0.0 and 10.0 based on the Pagefind docs (bottom of page): https://pagefind.app/docs/weighting/#ranking-content-higher-or-lower.

We could keep current behavior of `pagefind: boolean` in the front matter to prevent breaking changes by using a union for the Zod schema.

### Why is this feature necessary?

With large sites there may be specific pages you wish to rank higher than others (such as an introduction or overview page for a given concept before a page that contains error messages for that concept). 

### Do you have examples of this feature in other projects?

Algolia custom ranking: https://www.algolia.com/doc/guides/managing-results/must-do/custom-ranking/how-to/configure-custom-ranking/

### Participation

- [X] I am willing to submit a pull request for this feature.

**Comment by mrienstra** [comment]
A lot has changed since this issue was opened -- see https://starlight.astro.build/reference/configuration/#pagefind -- maybe this issue & PR can be closed...?

**Reply by delucis** [reply]
I guess there may still be interest at some point in **page-level** options as this is proposing.

Currently we only have a boolean frontmatter property that can be used to hide a page from search results: https://starlight.astro.build/reference/frontmatter/#pagefind
But no built-in support for boosting/deranking a page.

It is possible though via Pagefind’s [weighting feature](https://pagefind.app/docs/weighting/). At the most simple, this can even be done inside a content page:

```md
---
title: My unimportant page
---

<div data-pagefind-weight="0.1">

## About this page

Some more content

</div>
```

It’s quite out-of-date now, but there is a PoC PR from Lorenzo for how a frontmatter property might work: https://github.com/withastro/starlight/pull/2371
